src/OptionsResolver.php

```diff
@@ -0,0 +1,266 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sentry;
+
+use Psr\Log\LoggerInterface;
+
+class OptionsResolver
+{
+    /**
+     * Contains all default values and also acts as a kind of schema.
+     * Only values present in the defaults can be overwritten.
+     *
+     * @var array<string, mixed>
+     */
+    private $defaults = [];
+
+    /**
+     * List of allowed types for each top level array key.
+     *
+     * @var array<string, string[]>
+     */
+    private $allowedTypes = [];
+
+    /**
+     * List of valid values or a validation callback for each top level array key.
+     *
+     * @var array<string, mixed[]|callable>
+     */
+    private $allowedValues = [];
+
+    /**
+     * Stores normalizers for each top level array key. Normalizers are executed for defaults and user provided options.
+     *
+     * @var array<callable>
+     */
+    private $normalizers = [];
+
+    /**
+     * @param array<string, mixed> $defaults
+     */
+    public function setDefaults(array $defaults): void
+    {
+        $processed = [];
+
+        foreach ($defaults as $option => $defaultValue) {
+            [$isValid, $normalized] = $this->normalizeAndValidate($option, $defaultValue);
+
+            if (!$isValid) {
+                // Since defaults are used as fallback values if passed options are invalid, we want to
+                // get hard errors here to make sure we have something to fall back to.
+                throw new \InvalidArgumentException(\sprintf('Invalid default for option "%s"', $option));
```

Comment:
Good use case for an exception since this is an "us" problem and not caused by user input.